### PR TITLE
Hard gas limits browser form

### LINKED ISSUE #1155: Combine pending-tx and pending-tx-details components
They're logically coupled together, and were only ever separated for an old feature that we don't have anymore (rendering tx details in a browser notification).

Discussion on this PR:
https://github.com/MetaMask/metamask-plugin/pull/1154

### LINKED ISSUE #1221: Nonce / Gas price issue
I played around with WEI values while using Metamask in MKR.Market. Not I can't get any transaction across. I tried on a new browser with a fresh Metamask install. Still doesn't work.

I am getting: popup.js:84760 Transaction gas price is too low. There is another transaction with same nonce in the queue. Try increasing the gas price or incrementing the nonce.

### LINKED ISSUE #1230: Recalculate gas cost when manually adjusting
My initial gas cost was too high (I'm using browse solidity which auto sets it to the block limit). I tried manually adjusting downward, but the gas cost does not update.

https://ipfs.pics/QmayWCTeg6hRs19JESCmsZeTgBf23yFmds8MSkjDH5S4Z7

Max total is $0.98, and I have $2 in my account, but it won't let me submit the transaction

### LINKED ISSUE #1233: Don't mutate user input in gas fields
Default gas is like `2000000`.  If you highlight that `2` and delete it, before you can type `3` or something, the input is mutated to be simply `0`.

We shouldn't mutate user input.  We should simply interpret it as best as we can, giving warnings & errors accordingly.

### LINKED ISSUE #1234: Set hard lower limit for gas price
The cause of #1229 was that an early tx had a very low gas price set, and infura accepted it, but no miners did, and then all subsequent transactions had nonces incremented behind that transaction, essentially being blocked by it.

Could be easily fixed by enforcing hard lower limits for gas price.  Investigating valid values now.

### PR DESCRIPTION
**Pull request by danfinlay**
Fixes #1234 
Fixes #1233
Fixes #1230
Fixes #1155
Fixes #1221.

Normal browser form invalid messages don't work in an extension-popup context, so I've implemented some custom messages here.

Right now they stay up indefinitely, maybe we want them to disappear after a couple seconds.

Opening as a possible proposal.

Here I show both a single validation, and what it looks like if two things are outside the valid range.

screen shot 2017-03-22 at 3 03 45 pm
screen shot 2017-03-22 at 3 03 33 pm

### CODE REVIEW

**Comment by Zanibas** [comment]
So these errors are overlayed over the current interface?

It's good, but the styling needs to be worked on for sure.

**Comment by flyswatter** [comment]
Yeah they're absolutely positioned, so they just hover under the fields they refer to.

@VladTod would you maybe whip us up a better form validation message?

**Comment by VladTod** [comment]
Hello,  @flyswatter and @Zanibas !
Shouldn't the user have the chance to change the gas amount for example? to correct the error/warning.
We need to find a way to show the errors and let people change the field while the error is visible.
Maybe we could display them in the bottom? after the 3 buttons? and maybe just have an "*" next to the wrong field.

Will design it soon and show you what i mean

**Comment by flyswatter** [comment]
Maybe we should ship this first, style it later.

I'm open to all kinds of styling improvements, but I also want us to ship this ASAP, because we have new users every day who report having their entire accounts locked up because they don't have this feature.

The current way does let users edit the field.

Right now the message disappears once the value is correct.

We could *also* make it disappear after a time delay, or after being clicked.

I look forward to seeing what Vlad proposes, but I also want to get an MVP out the door soon.

**Comment by flyswatter** [comment]
Now also Fixes #1155.

File: ui/app/conf-tx.js [hunk 1]
```diff
@@ -172,47 +172,53 @@ ConfirmTxScreen.prototype.onTxChange = function (txData) {
...
   log.info('conf-tx.js: signing message')
   var params = msgData.msgParams
   params.metamaskId = msgData.id
-  event.stopPropagation()
+  this.stopPropagation(event)
   this.props.dispatch(actions.signMsg(params))
 }
 
+ConfirmTxScreen.prototype.stopPropagation = function (event) {
+  if (event.stopPropagation) {
```
Comment (kumavis) [inline ui/app/conf-tx.js]: whats this for? do some events not have this?

Comment (flyswatter) [inline ui/app/conf-tx.js]: Correct.  Specifically, the `invalid` event emitted by forms, so I just made this function so we could handle it generically with our other events in this component.

**Review by 2-am-zzz — CHANGES_REQUESTED** [review]
The reset button isn't functioning right now.

**Review by 2-am-zzz — APPROVED** [review]

**Comment by VladTod** [comment]
Hello Guys, here are two variations for the error "pop-up"
Important thing is that user should be able to edit the fields with the error visible. Otherwise people will forget what the error is, or what the number in the error is. For example on gas price error, 20000000 is hard to input without seeing it, because it has many 0's.

I'm not quite sure this first version is even possible since it's broken in 2 parts.
pop-uperror

And this is the second variation.
01-pop-uperror

Let me know what you think

**Pull request merged**